sockeye/loss.py

```diff
@@ -148,23 +191,58 @@ def get_loss(self, logits: mx.sym.Symbol, labels: mx.sym.Symbol) -> List[mx.sym.
-        off_value = self._alpha / (self._vocab_size - 1.0)
+        on_value = 1.0 - self.loss_config.smoothed_cross_entropy_alpha
+        off_value = self.loss_config.smoothed_cross_entropy_alpha / (self.loss_config.vocab_size - 1.0)
         cross_entropy = mx.sym.one_hot(indices=mx.sym.cast(data=labels, dtype='int32'),
-                                       depth=self._vocab_size,
+                                       depth=self.loss_config.vocab_size,
                                        on_value=on_value,
                                        off_value=off_value)
 
         # zero out pad symbols (0)
-        cross_entropy = mx.sym.where(labels, cross_entropy, mx.sym.zeros((0, self._vocab_size)))
+        cross_entropy = mx.sym.where(labels, cross_entropy, mx.sym.zeros((0, self.loss_config.vocab_size)))
 
         # compute cross_entropy
-        cross_entropy *= - mx.sym.log(data=probs + 1e-10)
+        cross_entropy = cross_entropy * - mx.sym.log(data=probs + 1e-10)
         cross_entropy = mx.sym.sum(data=cross_entropy, axis=1)
 
-        if self._normalize:
+        if self.loss_config.normalize:
             cross_entropy = _normalize(cross_entropy, labels)
 
         cross_entropy = mx.sym.MakeLoss(cross_entropy, name=C.SMOOTHED_CROSS_ENTROPY)
         probs = mx.sym.BlockGrad(probs, name=C.SOFTMAX_NAME)
         return [cross_entropy, probs]
+
+    def create_metric(self) -> "SmoothedCrossEntropyMetric":
+        return SmoothedCrossEntropyMetric(self.loss_config)
+
+
+class SmoothedCrossEntropyMetric(EvalMetric):
+    """
+    Metric wrapper for smoothed cross entropy loss.  Since SCE already returns loss values during
+    training, this class simply sums the results.
+
+    :param loss_config: The configuration used for the corresponding loss.
+    :param name: Name of this metric instance for display.
+    :param output_names: Name of predictions that should be used when updating with update_dict.
+    :param output_labels: Name of labels that should be used when updating with update_dict.
+    """
+    def __init__(self,
+                 loss_config: LossConfig,
+                 name: str = C.SMOOTHED_CROSS_ENTROPY,
+                 output_names: Optional[List[str]] = None,
+                 label_names: Optional[List[str]] = None) -> None:
+        super().__init__(name, output_names=output_names, label_names=label_names)
+        self.loss_config = loss_config
+
+    def update(self, labels, preds):
+        # Take cross entropy values only
+        sces = preds[::2]
+        for label, sce in zip(labels, sces):
+            # SCE is pre-computed, so just sum
+            self.sum_metric += sce.asnumpy().sum()
+            # Only scale if loss is not already normalized
+            if self.loss_config.normalize:
+                self.num_inst += 1
+            else:
+                ignore = (label == C.PAD_ID).astype(dtype=sce.dtype)
+                self.num_inst += label.size - mx.nd.sum(ignore).asscalar()
```

Comment:
same thing here. If `normalize=False` we should normalize by the batch size. This is btw done through the `rescale_grad` argument to the optimizer in [train.py](https://github.com/awslabs/sockeye/blob/master/sockeye/train.py#L533). We set this argument explicitly, but it is also what MXNet does by default.

Reply:
Same as above.  Right now we're doing the same thing in a few different ways and a few different places.  It would be great if MXNet's loss and metrics were more synchronized.

Reply:
Well in our case we know that with `normalize=False` the loss is something like:
```
loss = 1./batch_size * sum(loss(next_word) for next_word in batch)
```
and I think it would make sense to define the metric accordingly.

Reply:
I see what you're saying now.  With `normalize=False`, we divide by batch size.  MXNet metrics divide by number of elements, sometimes ignoring padding.  Right now we're consistent with metrics whereas if we're using this as training feedback, we want to be consistent with loss.  Makes sense, though it also sounds like the best way to avoid problems is to always use `normalize=True` : )  I'll update and add some comments to this effect.